devito/operations/solve.py

```diff
@@ -0,0 +1,120 @@
+from functools import singledispatch
+
+import sympy
+
+from devito.logger import warning
+from devito.finite_differences.differentiable import Add, Mul, EvalDerivative
+from devito.finite_differences.derivative import Derivative
+from devito.tools import as_tuple
+
+__all__ = ['solve', 'linsolve']
+
+
+class SolveError(Exception):
+    """Base class for exceptions in this module."""
+    pass
+
+
+def solve(eq, target, **kwargs):
+    """
+    Algebraically rearrange an Eq w.r.t. a given symbol.
+
+    This is a wrapper around ``sympy.solve``.
+
+    Parameters
+    ----------
+    eq : expr-like
+        The equation to be rearranged.
+    target : symbol
+        The symbol w.r.t. which the equation is rearranged. May be a `Function`
+        or any other symbolic object.
+    **kwargs
+        Symbolic optimizations applied while rearranging the equation. For more
+        information. refer to ``sympy.solve.__doc__``.
+    """
+
+    if hasattr(eq, 'rhs'):
+        eq = eq.lhs - eq.rhs if eq.rhs != 0 else eq.lhs
+    sols = []
+    for e, t in zip(as_tuple(eq), as_tuple(target)):
+        # Try first linear solver
+        try:
+            sols.append(linsolve(eval_t(e), t))
+        except SolveError:
+            warning("Equation is not affine w.r.t the target, falling back to standard"
+                    "sympy.solve that may be slow")
+            kwargs['rational'] = False  # Avoid float indices
+            kwargs['simplify'] = False  # Do not attempt premature optimisation
+            sols.append(sympy.solve(e.evaluate, t, **kwargs)[0])
+    # We need to rebuild the vector/tensor as sympy.solve outputs a tuple of solutions
+    if len(sols) > 1:
+        return target.new_from_mat(sols)
+    else:
+        return sols[0]
```

Comment:
I think, though I may be wrong, that this would fail "weirdly" if one accidentally called `solve([], ...` ie with no equations. You probably want to have an as_tuple at the very beginning of this function, and sanity-check that `len(...) > 0` , so surely `return sols[0]` won't trigger an `IndexError`